### ISSUE TITLE
how does the quality option of im2rec affect?

### ISSUE THREAD

**Issue by zlmzju** [tornadomeet/ResNet#17]
Luckily I have found this great work! I reproduced the ResNet-v1 on ImageNet a few months ago (50-layer top5 error: 7.4), and now I realize that the images are encoded with the default quality option `quality=80`. 

I use the python tool `im2rec.py` on Windows. And the `quality=80` affects a lot when I train another network on Cifar100 because the `32x32` images are too small. 
1. So  I wonder if I have to retrain my own network on ImageNet by using a high quality `record file`, such as `quality=90 or 100`? How does this option affect, 0.1 percent difference?
2. Another question on `im2rec`. Now I convert the Cifar100 python version data to lossless `png` images, and pack them into record file by `im2rec.py`, am I right? Do you guys know how the provided Cifar-10 record file generated? (Do I need to compile the `im2rec.cc` on Windows?)

I would greatly appreciate it if you kindly give me some feedback.

**Comment by tornadomeet** [comment]
hello, @zlmzju 
1. i think you do not need to retrain with `quality=90`, because your resnet-50 has achieved the result in resent-v1 paper.  but i think `quality=90` will lead better result than `quality-80`, should be about ~0.5 on top-1.
2.  you get the right way on convert cifar100, and no need using im2rec.cc, because im2rec.py should get the same reulst(but i don't used it before). i just download the existed rec of cifar10(the quality should be 80). another, i suggest using `quality=100` when convert cifar10 and cifar100, because the size is not so big.

BTW, i trained one model of resnet-50 on imagnet using quality=80 before,  but the result is about 7.9(<100 epoch), and your result looks good than mine, so would you describe some details which is  different with this repo?

**Comment by zlmzju** [comment]
Thanks for your reply, @tornadomeet 
1. The ~0.5 improvement sounds good, and I will use `quality>=90` in my future experiments.
2. Now I achieve normal performance on cifar100 by using the `png` format, thanks for your explanation.
3. I have read your codes, and find that the data normalization method and training strategies are different. 
   - Instead of using a `BN` layer, I simply use the `mean_r, mean_g, mean_b` and `scale` values to normalize the input data. 
   - Besides that, I trained the model for 90 epochs, and the learning rate is reduced by a factor of 2 every 10 epochs. You suggest that it's better for running longer than 30 epoch before first decrease, and  I agree because I observed that lately reducing the `LR` is better for deep networks on Cifar10. So maybe this difference does not matter.
   - The data augmentation methods are similar, and I try to reproduce every details in the [torch repo](https://github.com/facebook/fb.resnet.torch), but I find it's difficult because I am not familar with mxnet. I previously achieved a similar result by running the torch code with the same training strategy (lr*0.5 every 10 epochs). (torch achieved 7.5 on top5 for 75 epochs, then I have to move from Linux to Windows for some reason) 

About the data augmentation, why do you cancel the scale/color/aspect augmentation during training after 95 epochs?

**Comment by tornadomeet** [comment]
thanks @zlmzju 
when training near the end, cancel scale/color/aspect will leads about 1.0~1.5 better result(for all resnet-18/34/50/101/152), you can check the log in https://github.com/tornadomeet/ResNet/tree/master/log.  
the reason is that i found that train error is higher than val even if we trained the model about 100 epochs, this is a strange! and it due to that we use shorted edge=480 in the rec, and random resize it to (255, 480), this is a big augmentation since we only crop 224*224 size in the training. so at last we'd better cancel this aug and let is to be close to the val.

the same augmentation io in torch will come in mxnet recentlly, r.f.https://github.com/dmlc/mxnet/pull/3227  

ps, would you provide the converted rec of cifar100? i'll put it on http://data.dmlc.ml/mxnet/data/, this is convenient for others do cifar-100 experiment.

**Comment by zlmzju** [comment]
Interesting! The data processing really matters in many practical scenarios. The python based ImageIter and Augumenter are exactly what I need, thanks for your efforts @tornadomeet . 

I have put my cifar100 rec files on [google drive (cifar100.zip, 125MB)](https://drive.google.com/open?id=0By55MQnF3PHCTHR5Y3FIRjdTZUE). 
- I convert the original images to this rec files without compression by using the lossless `png` format.  
- I compute the mean file by mxnet using the converted rec files, and the final mean values are `(mean_r, mean_g, mean_b)=(129.304, 124.070, 112.434)`, which is exactly the same as the original python data's mean. BTW, the std values are `[68.170,  65.392  70.418]`.

**Comment by tornadomeet** [comment]
@zlmzju  Thanks!

**Comment by shipengai** [comment]
@zlmzju How to compute the mean file by mxnet using the converted rec files? Thanks
